### PR TITLE
Feature #18197: Allow GA4 tracking

### PR DESCRIPTION
**Pull request by gabrieljenik**

### CODE REVIEW

File: themes/survey/vanilla/views/subviews/header/google_analytics.twig [hunk 1]
```diff
@@ -15,26 +15,46 @@
...
 #}
 {% if aSurveyInfo.googleanalyticsapikey  is defined and aSurveyInfo.googleanalyticsapikey and aSurveyInfo.googleanalyticsstyle %}
-    {% if aSurveyInfo.googleanalyticsstyle  is defined and aSurveyInfo.googleanalyticsstyle == 1 %}
-        <script>
-        (function(i,s,o,g,r,a,m){ i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
-            (i[r].q=i[r].q||[]).push(arguments) },i[r].l=1*new Date();a=s.createElement(o),
-            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
-        })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
+    {# Check if the tracking ID is a Universal Analytics (legacy) or a Google Analytics 4 ID #}
+    {% if aSurveyInfo.googleanalyticsapikey starts with 'UA-' %}
```
Comment (Shnoulle) [inline themes/survey/vanilla/views/subviews/header/google_analytics.twig:19]: :+1:

**Review by Shnoulle — APPROVED** [review]
I don't check if it's OK with GA, only check LimeSurvey part

File: themes/survey/vanilla/views/subviews/header/google_analytics.twig [hunk 2]
```diff
@@ -15,26 +15,46 @@
...
+            ga('create', '{{ aSurveyInfo.googleanalyticsapikey }}', 'auto');
+            ga('send', 'pageview');
+            ga('send', 'pageview', '{{ trackUrl }}');
+            </script>
+        {% endif %}
     {% else %}
-        {% set trackUrl = getGoogleAnalyticsTrackingUrl(aSurveyInfo.sid, aSurveyInfo.trackUrlPageName ?? '') %}
+        {# GA4 #}
+        <!-- Google tag (gtag.js) -->
+        <script async src="https://www.googletagmanager.com/gtag/js?id={{aSurveyInfo.googleanalyticsapikey}}"></script>
```
Comment (Shnoulle) [inline themes/survey/vanilla/views/subviews/header/google_analytics.twig:46]: We don't encode googleanalyticsapikey in controller, and don't filter (or fix) in model rules

I 'm sure i can find XSS here.

Comment (Shnoulle) [reply themes/survey/vanilla/views/subviews/header/google_analytics.twig:46]: Not same method : now it's on an url, before in a JS var.

File: themes/survey/vanilla/views/subviews/header/google_analytics.twig (hunk 1 above)
Comment (maziminke) [reply themes/survey/vanilla/views/subviews/header/google_analytics.twig:19]: @Shnoulle @gabrieljenik can we please merge this into the next 3.x release? We would love to use this for a new project starting next week.

**Comment by gabrieljenik** [comment]
Tested this.

UA:
image

GA4:
image

Haven't checked looking at this on the GA side.

**Pull request merged**